File: Algorithms/Easy/RemoveDuplicatesFromSortedArray.go
package easy

// Time complexity: O(n)
// Space complexity: O(n)
func removeDuplicates_1(nums []int) int {
	seen := make(map[int]struct{})
	idx := 0
	for _, num := range nums {
		if _, ok := seen[num]; !ok {
			nums[idx] = num
			idx++
			seen[num] = struct{}{}
		}
	}

	return len(seen)
}

// Time complexity: O(n)
// Space complexity: O(1)
func removeDuplicates_2(nums []int) int {
	idx := 1
	for i := 1; i < len(nums); i++ {
		if nums[i] != nums[i-1] {
			nums[idx] = nums[i]
			idx++
		}
	}

	return idx
}

File: Algorithms/Easy/26.Remove-Duplicates-From-Sorted-Array/RemoveDuplicatesFromSortedArray.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
    def removeDuplicates_1(self, nums: list[int]) -> int:
        seen = set()
        lastUnique = 0
        for num in nums:
            if num not in seen:
                seen.add(num)
                nums[lastUnique] = num
                lastUnique += 1

        return len(seen)

    # Time complexity: O(n)
    # Space complexity: O(1)
    def removeDuplicates_2(self, nums: list[int]) -> int:
        l = 1
        for r in range(1, len(nums)):
            if nums[r] != nums[r - 1]:
                nums[l] = nums[r]
                l += 1

        return l
